src/bind.c

```diff
@@ -122,3 +122,175 @@ static SEXP as_df_row_impl(SEXP x, bool quiet) {
+
+// [[ register(external = TRUE) ]]
+SEXP vctrs_cbind(SEXP call, SEXP op, SEXP args, SEXP env) {
+  args = CDR(args);
+
+  SEXP xs = PROTECT(rlang_env_dots_list(env));
+  SEXP ptype = PROTECT(Rf_eval(CAR(args), env));
+  SEXP size = PROTECT(Rf_eval(CADR(args), env));
+
+  SEXP out = vec_cbind(xs, ptype, size);
+
+  UNPROTECT(3);
+  return out;
+}
+
+static SEXP vec_cbind(SEXP xs, SEXP ptype, SEXP size) {
+  R_len_t n = Rf_length(xs);
+
+  // Find the common container type of inputs
+  SEXP containers = PROTECT(map(xs, &cbind_container_type));
+  ptype = PROTECT(cbind_container_type(ptype));
+
+  SEXP type = PROTECT(vctrs_type_common_impl(containers, ptype));
+  if (type == R_NilValue) {
+    type = new_data_frame(vctrs_shared_empty_list, 0);
+  } else if (!is_data_frame(type)) {
+    type = r_as_data_frame(type);
+  }
+  UNPROTECT(1);
+  PROTECT(type);
+
+
+  R_len_t nrow;
+  if (size == R_NilValue) {
+    nrow = vec_size_common(xs);
+  } else {
+    nrow = size_validate(size, ".size");
+  }
+
+
+  // Validate input data frames and collect total number of columns
+  SEXP xs_names = PROTECT(r_names(xs));
+  bool has_names = xs_names != R_NilValue;
+  SEXP* xs_names_ptr = has_names ? STRING_PTR(xs_names) : NULL;
+
+  R_len_t ncol = 0;
+  for (R_len_t i = 0; i < n; ++i, ++xs_names_ptr) {
+    SEXP x = VECTOR_ELT(xs, i);
+
+    if (x == R_NilValue) {
+      continue;
+    }
+
+    x = PROTECT(vec_recycle(x, nrow));
+    x = PROTECT(as_df_col(x, has_names ? *xs_names_ptr : R_NilValue));
+
+    SET_VECTOR_ELT(xs, i, x);
+    UNPROTECT(2);
+
+    ncol += Rf_length(x);
+  }
+
+
+  // Fill in columns
+  SEXP out = PROTECT(Rf_allocVector(VECSXP, ncol));
+  SEXP names = PROTECT(Rf_allocVector(STRSXP, ncol));
+
+  SEXP idx = PROTECT(compact_seq(0, 0));
+  int* idx_ptr = INTEGER(idx);
+
+  R_len_t counter = 0;
+
+  for (R_len_t i = 0; i < n; ++i) {
+    SEXP x = VECTOR_ELT(xs, i);
+
+    if (x == R_NilValue) {
+      continue;
+    }
+
+    R_len_t xn = Rf_length(x);
+    idx_ptr[0] = counter;
+    idx_ptr[1] = counter + xn;
+
+    list_assign(out, idx, x, false);
+
+    SEXP xnms = PROTECT(r_names(x));
+    if (xnms != R_NilValue) {
+      chr_assign(names, idx, xnms, false);
+    }
+
+    counter += xn;
+    UNPROTECT(1);
+  }
+
+  names = PROTECT(as_unique_names(names, false));
+  Rf_setAttrib(out, R_NamesSymbol, names);
+
+  out = vec_restore(out, type, R_NilValue);
+
+  UNPROTECT(8);
+  return out;
+}
+
+static SEXP cbind_container_type(SEXP x) {
+  if (is_data_frame(x)) {
+    return df_container_type(x);
+  } else {
+    return R_NilValue;
+  }
+}
+
+
+static SEXP df_as_df_col(SEXP x, SEXP outer);
+static SEXP shaped_as_df_col(SEXP x, SEXP outer);
+static SEXP vec_as_df_col(SEXP x, SEXP outer);
+
+// [[ register() ]]
+SEXP vctrs_as_df_col(SEXP x, SEXP outer) {
+  return as_df_col(x, r_chr_get(outer, 0));
+}
+static SEXP as_df_col(SEXP x, SEXP outer) {
+  if (is_data_frame(x)) {
+    return df_as_df_col(x, outer);
+  }
```

Comment:
Feels weird to use early-return style here. I think it would be clearer as a single if block.

Reply:
Ah that's because I find chained if-else very hard to read. With early return style, it's easier for me to check the last expression of each block and determine that the control flow pattern is homogeneous.